Documentation/Evolution/CharacterClasses.md

```diff
@@ -0,0 +1,354 @@
+| `[:alnum:]`  | `[A-Za-z0-9]`     |
+| `[:alpha:]`  | `[A-Za-z]`        |
+| `[:ascii:]`  | `[\x00-\x7F]`     |
+| `[:blank:]`  | `[ \t]`           |
+| `[:cntrl:]`  | `[\x00-\x1F\x7F]` |
+| `[:digit:]`  | `[0-9]`           |
+| `[:graph:]`  | `[\x21-\x7E]`     |
+| `[:lower:]`  | `[a-z]`           |
+| `[:print:]`  | `[[:graph:] ]`    |
+| `[:punct:]`  | `[-!"#$%&'()*+,./:;<=>?@[\\\]^_{|}~]` |
+| `[:space:]`  | `[ \t\n\r\f\v]`   |
+| `[:upper:]`  | `[A-Z]`           |
+| `[:word:]`   | `[A-Za-z0-9_]`    |
+| `[:xdigit:]` | `[0-9A-Fa-f]`     |
+
+### Custom Classes: `[...]`
+
+Users can create custom character classes by using literal characters and most predefined character classes within square brackets.
+
+- Individual characters and character classes described above can be used as is.
+- Ranges of Unicode scalar values can be specified by separating the start and end of the range (inclusive) with a hyphen. Note that because `Character` isn't a `Comparable` type, ranges can only match Unicode scalar values. Crossing the UTF-16 surrogate pair range (U+D800 to U+DFFF) is permitted, but values in that range do not match anything, regardless of the string's encoding.
+- Include a literal hyphen by escaping, or by placing it first or last in the custom class.
+- Negate a custom class by placing a caret immediately after the opening bracket (e.g. `[^a-zA-Z]`).
+
+
+## Alternatives Considered
+
+### Applying Regular Expressions to String Views
+
+A prior design allowed a user to choose their desired matching level and semantics by applying a regular expression to a string's `UnicodeScalarView`, `UTF16View`, or `UTF8View`. While initially appealing, this approach posed multiple problems:
+
+- Since we want regular expressions created with a literal to be usable all the same places as those created using the DSL, anything predicate-based needs to have the input type specified at the time of creation. It wouldn't make sense to add a `(Character) -> Bool` predicate to a regular expression, and then apply it to the string's `UTF8View`.
+- The matching semantics of a regular expression are closely tied to the way it's composed, so it would likely be unpredictable or even nonsensical to apply a regular expression written for one view of a string to be applied to another view.
+- It's unclear whether the two UTF-encoded views would treat their elements as individual `UInt16` or `UInt8` values, or whether they would retain some notion of being encoded Unicode data. If the former, how would character classes like `\u{...}` or `\p{...}` be used by the parser? And if the latter, what purpose does this serve beyond parsing the `UnicodeScalarView`?
+
+For these reasons, regular expressions will target only the `StringProtocol`-conforming types — `String` and `Substring`. The other string views can be parsed using the more general `Collection`-based pattern matching, without the regular expression-specific features described in this proposal.
+
```

Comment:
We can say that API discussion, how modes switch, etc., are out of scope of this semantics-clarifying pitch. But I think it's suspect to make this decision in a pitch when corresponding API doesn't exist yet nor is being pitched or discussed here.